accepted/2021/statics-in-interfaces/README.md

```diff
@@ -6,8 +6,1267 @@
+
+        static abstract TResult operator %(TSelf lhs, TOther rhs);
+
+        static abstract TResult checked operator %(TSelf lhs, TOther rhs);
+    }
+
+    public interface INegatable<TSelf, TResult>
+        where TSelf : INegatable<TSelf, TResult>
+    {
+        // Should unary plus be on its own type?
+
+        static abstract TResult operator +(TSelf value);
+
+        static abstract TResult checked operator +(TSelf value);
+
+        static abstract TResult operator -(TSelf value);
+
+        static abstract TResult checked operator -(TSelf value);
+    }
+
+    public interface IFixedWidth<TSelf>
+        where TSelf : IFixedWidth<TSelf>
+    {
+        // MinValue and MaxValue are more general purpose than just "numbers", so they live on this common base type
+        // Fixed-Width isn't the best name however, so we should think of alternatives
+
+        static abstract TSelf MinValue { get; }
+
+        static abstract TSelf MaxValue { get; }
+    }
+}
+```
+
+### Numeric Interfaces
+
+The numeric interfaces build upon the base interfaces by defining the core abstraction that most users are expected to interact with.
+
+```csharp
+namespace System
+{
+    public interface INumber<TSelf>
+        : IAddable<TSelf, TSelf>,
+          IComparableOperators<TSelf, TSelf>,   // implies IEquatableOperators<TSelf, TSelf>
+          IDecrementable<TSelf>,
+          IDivisable<TSelf, TSelf>,
+          IIncrementable<TSelf>,
+          IMultipliable<TSelf, TSelf>,
+          IRemainder<TSelf, TSelf>,
+          ISpanFormattable,                     // implies IFormattable
+          ISpanParseable<TSelf>,                // implies IParseable<TSelf>
+          ISubtractable<TSelf, TSelf>
+        where TSelf : INumber<TSelf>
+    {
+        // For the Create methods, there is some concern over users implementing them. It is not necessarily trivial to take an arbitrary TOther
+        // and convert it into an arbitrary TSelf. If you know the type of TOther, you may be able to optimize in a few ways. Otherwise, you
+        // may have to fail and throw in the worst case.
+
+        static abstract TSelf Create<TOther>(TOther value)
+            where TOther : INumber<TOther>;
+
+        static abstract TSelf? CreateExactly<TOther>(TOther value)
+            where TOther : INumber<TOther>;
+
+        static abstract TSelf CreateSaturating<TOther>(TOther value)
+            where TOther : INumber<TOther>;
+
+        static abstract TSelf CreateTruncating<TOther>(TOther value)
+            where TOther : INumber<TOther>;
+
+        // There is an open question on whether properties like IsSigned, IsBinary, IsFixedWidth, Base/Radix, and others are beneficial
+        // We could expose them for trivial checks or users could be required to check for the corresponding correct interfaces
+
+        static abstract TSelf One { get; }
+
+        static abstract TSelf Zero { get; }
```

Comment:
Would it be possible to move these up to some of the parent interfaces? For example, move `Zero` to `IAddable<T>` and move `One` to `IMultipliable<T>`?

There are mathematical types (for example Matrices or Elliptic Curves) where they have identity elements for the operations they support, but they don't have support for additional operations (ex. Matrices doesn't support Division but has a Zero and One. Elliptic Curves have a Zero and addition but don't have multiplication or a One element)

Reply:
Not easily. You have to consider whether it should be `TSelf One/Zero` or `TOther One/Zero` or `TResult One/Zero`. Likewise, not everything has a "clear" concept of `one` vs `zero`. It also gets problematic when considering other types like `ISubtractable` or `IDivisible` as well as  when considering types that implement multiple `IAddable` or `IMultipliable` interfaces (differing on `TOther` at least), The latter of which requires you to consider if all concepts of `One` are equivalent.

Instead, my current thinking is that there will need to be a different interface for "sets of numbers" (which includes vectors, matrices, tensors, etc). This might simply be `IVector<T>`, `IMatrix<T>`, and `ITensor<T>`, but its something I'm actively trying to get a design for.

Reply:
Yes, the identity element only makes sense for closed operations where `TSelf`, `TOther`, and `TResult`, are all the same type. So they'd require a one-parameter version of `IAddable<T>` and `IMulitpliable<T>` to make sense.

Even if there were other interfaces like `IAddableIdentity<T>` and `IMultipliableIdentity<T>`, I'd be happy with that. I want to be able to write generic math that works at a group/ring level and I can't do that if the identity element property is specific to numbers, or vectors, or matrices, or curves, etc.

Reply:
I don't think we are going to be exposing concepts like `group`/`ring` or `monoid`.

We must expose these concepts via interfaces, not traits, and there are far too many existing .NET types and concepts that do not fit the mathematical model you find in many functional languages, like Haskell.

There is also a limit on the number of interfaces we can add here. The number of interfaces currently modeled here is already "a lot" and is expected to have a measurable impact to application startup.

Limitations in the runtime also come into play here:
* We can't overload based on generic constraint
* You can only use these via constrained generics (because these must be a `constrained call`)
* Changing your generic constraint is a breaking change
* There is no concept of associated types
* etc

All of these add up to defining things more like `IComparable` or `IEquatable`, which expose a general contract for a given operation but don't provide more than a few guarantees and so `IAddable` will provide a base interface for building higher abstractions, like `INumber` or `IVector` and most code will need to take types like `INumber` or `IVector` and some very specialized methods can take a base interface like `IAddable`.

Some concepts like `Commutativity`, `Associativity`, or `Distributivity` might be possible to light up for tools via attributes. It will likely not be a first class citizen due to JIT time constraints and cases like `float` where most of the rules don't exist in IEEE 754 compliant code.

Reply:
We will certainly discuss this in API review and I will be raising what might be left on the table if we don't expose these concepts in v1.

But as of right now, these concepts are not trivial to model for C#-like .NET code, which will be the primary user of these

Reply:
One example to definitely think about is that you can't write a `Sum` function for `IEnumerable<IAddable<T, U, V>>` without either having an identity element on the interface or requiring users to pass one in.

Reply:
After a lot of thought, I added `IAdditiveIdentity` and `IMultiplicativeIdentity` to fill this gap as I believe its the simplest way to cover it and fit in with the existing model.